pkg/service/endpoints_controller.go

```diff
@@ -26,135 +27,280 @@ import (
+			},
+		},
+		&api.Service{},
+		FullServiceResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: e.enqueueService,
+			UpdateFunc: func(old, cur interface{}) {
+				// TODO: change old & new services!
+				e.enqueueService(cur)
+			},
+			DeleteFunc: e.enqueueService,
+		},
+	)
+
+	e.podStore.Store, e.podController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return e.client.Pods(api.NamespaceAll).List(labels.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return e.client.Pods(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.Pod{},
+		5*time.Minute,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: e.addPodServicesToQueue,
+			UpdateFunc: func(old, cur interface{}) {
+				if api.Semantic.DeepEqual(old, cur) {
+					return
+				}
+				e.addPodServicesToQueue(cur)
+			},
+			DeleteFunc: e.deletePod,
+		},
+	)
+
+	return e
+}
+
 // EndpointController manages selector-based service endpoints.
 type EndpointController struct {
 	client *client.Client
+
+	serviceStore cache.StoreToServiceLister
+	podStore     cache.StoreToPodLister
+
+	// Services that need to be updated. A channel is inappropriate here,
+	// because it allows services with lots of pods to be serviced much
+	// more often than services with few pods; it also would cause a
+	// service that's inserted multiple times to be processed more than
+	// necessary.
+	queue *workqueue.Type
+
+	// Since we join two objects, we'll watch both of them with
+	// controllers.
+	serviceController *framework.Controller
+	podController     *framework.Controller
 }
 
-// NewEndpointController returns a new *EndpointController.
-func NewEndpointController(client *client.Client) *EndpointController {
-	return &EndpointController{
-		client: client,
+// Runs e; will not return until stopCh is closed. workers determines how many
+// endpoints will be handled in parallel.
+func (e *EndpointController) Run(workers int, stopCh <-chan struct{}) {
+	defer util.HandleCrash()
+	go e.serviceController.Run(stopCh)
+	go e.podController.Run(stopCh)
+	for i := 0; i < workers; i++ {
+		go util.Until(e.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	e.queue.ShutDown()
+}
+
+func (e *EndpointController) deletePod(obj interface{}) {
+	if _, ok := obj.(*api.Pod); ok {
+		e.addPodServicesToQueue(obj)
+		return
+	}
+	podKey, err := framework.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+	}
+	glog.Infof("Pod %q was deleted but we don't have a record of its final state, so it will take up to %v before it will be removed from all endpoint records.", podKey, FullServiceResyncPeriod)
+
+	// TODO: keep a map of pods to services to handle this condition.
+}
+
+func (e *EndpointController) addPodServicesToQueue(obj interface{}) {
```

Comment:
AFAIU this function adds to queu services that this pod belongs to. Right? Can you please add a comment?